cpp/src/arrow/compute/exec/options.h

```diff
@@ -77,6 +77,34 @@ class ARROW_EXPORT TableSourceNodeOptions : public ExecNodeOptions {
   int64_t max_batch_size;
 };
 
+/// \brief An extended Source node which accepts a schema
+///
+/// ItMaker is a maker of an iterator of tabular data.
```

Comment:
Why are we taking in a factory?  It seems that `it_maker` is called exactly once, when the node is created.  So what is the advantage to accepting, for example, `std::function<Iterator<std::shared_ptr<ExecBatch>>()>` instead of `Iterator<std::shared_ptr<ExecBatch>>`?

Reply:
My (hopefully not wrong) understanding is that options should have value semantics, so that they could be safely copied or passed multiple times as an argument. A factory preserves this semantics whereas an iterator doesn't, because it is stateful.